File: CollectionLayout/ClassGroup/ViewGroup/CollectionCardCell.m
//
//  CollectionCardCell.m
//  CollectionLayout
//
//  Created by 苏沫离 on 2020/8/11.
//  Copyright © 2020 苏沫离. All rights reserved.
//

#import "CollectionCardCell.h"

@interface CollectionCardCell ()
@property (nonatomic ,strong) UIImageView *imageView;
@property (nonatomic ,strong) UILabel *nameLable;
@end
@implementation CollectionCardCell

- (instancetype)initWithFrame:(CGRect)frame{
    self = [super initWithFrame:frame];
    if (self){
        self.clipsToBounds = YES;
        self.backgroundColor = UIColor.whiteColor;
        [self.contentView addSubview:self.imageView];
        [self.contentView addSubview:self.nameLable];
    }
    return self;
}

- (void)layoutSubviews{
    [super layoutSubviews];
    CGFloat contentWidth = CGRectGetWidth(self.contentView.bounds);
    self.imageView.frame = CGRectMake(0, 0, contentWidth, contentWidth);
    self.nameLable.frame = CGRectMake(0, CGRectGetMaxY(self.imageView.frame), contentWidth, 40.0);
}

#pragma mark - public method

- (void)setModel:(DataModel *)model{
    _model = model;
    self.imageView.image = model.image;
    self.nameLable.text = [NSString stringWithFormat:@"%@-%ld",model.title,model.index];
}

#pragma mark - getter and setter

- (UIImageView *)imageView{
    if (_imageView == nil) {
        _imageView = [[UIImageView alloc] initWithFrame:CGRectMake(12, 20, 42, 42)];
        _imageView.backgroundColor = [UIColor colorWithRed:222/255.0 green:222/255.0 blue:222/255.0 alpha:1.0];
        _imageView.contentMode = UIViewContentModeScaleAspectFill;
        _imageView.clipsToBounds = YES;
    }
    return _imageView;
}

- (UILabel *)nameLable{
    if (_nameLable == nil){
        UILabel *label = [[UILabel alloc] init];
        label.backgroundColor = [UIColor colorWithPatternImage:[UIImage imageNamed:@"cardItemBack"]];
        label.textAlignment = NSTextAlignmentCenter;
        label.textColor = [UIColor whiteColor];
        label.font = [UIFont fontWithName:@"Helvetica-Bold" size:20];
        _nameLable = label;
    }
    return _nameLable;
}

@end

File: CollectionLayout-Swift/CollectionLayout-Swift/ControllerGroup/ViewGroup/CollectionCardCell.swift
//
//  CollectionCardCell.swift
//  CollectionLayout-Swift
//
//  Created by 苏沫离 on 2020/9/11.
//  Copyright © 2020 苏沫离. All rights reserved.
//

import UIKit

class CollectionCardCell: UICollectionViewCell {
    var nameLable: UILabel {
        get {
            let label = UILabel()
            label.backgroundColor = UIColor(patternImage: UIImage(named: "cardItemBack")!)
            label.textAlignment = .center
            label.textColor = UIColor.white
            label.font = UIFont.init(name: "Helvetica-Bold", size: 20)
            return label
        }
    }
    
    var imageView: UIImageView {
        get {
            let imageV = UIImageView(frame: CGRect(x: 12, y: 20, width: 42, height: 42))
            imageV.backgroundColor = UIColor(red: 222/255.0, green: 222/255.0, blue: 222/255.0, alpha: 1)
            imageV.contentMode = .scaleAspectFill
            imageV.clipsToBounds = true
            return imageV
        }
    }
    
    var model: DataModel! {
        didSet{
            imageView.image = model.image
            nameLable.text = "\(model.title ?? "")-\(model.index ?? 0)"
        }
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.clipsToBounds = true
        self.backgroundColor = UIColor.white
        self.contentView.addSubview(imageView)
        self.contentView.addSubview(nameLable)
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        let contentWidth = self.contentView.bounds.width
        imageView.frame = CGRect(x: 0, y: 0, width: contentWidth, height: contentWidth)
        nameLable.frame = CGRect(x: 0, y: imageView.frame.maxY, width: contentWidth, height: 40)
    }
    
    override func apply(_ layoutAttributes: UICollectionViewLayoutAttributes) {
        if layoutAttributes.isKind(of: YLCollectionReelLayoutAttributes.self) {
            let reelAttributes = layoutAttributes as! YLCollectionReelLayoutAttributes
            self.layer.anchorPoint = reelAttributes.anchorPoint
            let y : CGFloat = self.layer.position.y + (reelAttributes.anchorPoint.y - 0.5) * self.bounds.size.height
            self.layer.position = CGPoint(x: self.layer.position.x, y: y)
        }
    }
        
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
}
